### PR TITLE
BTreeMap: tougher checking on most uses of copy_nonoverlapping 

### PR DESCRIPTION
**Pull request by ssomers**
Miri checks pointer provenance and destination, but we can check it in debug builds already.
Also, we can let Miri confirm we don't mistake imprints of moved keys and values as genuine.
r? @Mark-Simulacrum

### CODE REVIEW

**Comment by ssomers** [comment]
Includes #80390. I only just noticed that many slice functions like `copy_from_slice` are defined as assignment: destination before source. `slice_move` is like `ptr::copy_nonoverlapping`: source then destination. Which one is best?

**Comment by Mark-Simulacrum** [comment]
I don't think order matters too much, but we can switch to using copy_nonoverlapping_to which is defined directly on raw pointers, which seems much better to me than thinking about order of arguments :)

Ideally I think we'd seek to do the same for the slice functions, and we might look at putting them on MaybeUninit itself.

File: library/alloc/src/collections/btree/node.rs [hunk 1]
```diff
@@ -1741,5 +1729,27 @@ unsafe fn slice_remove<T>(slice: &mut [MaybeUninit<T>], idx: usize) -> T {
...
+///
+/// # Safety
+/// The `dest` slice is at least as long as the `source` slice.
+unsafe fn slice_move<T>(source: &mut [MaybeUninit<T>], dest: &mut [MaybeUninit<T>]) {
+    debug_assert!(dest.len() >= source.len());
+    unsafe {
+        ptr::copy_nonoverlapping(source.as_ptr(), dest.as_mut_ptr(), source.len());
+    }
+    if cfg!(debug_assertions) {
+        // Tell Miri that the copied slice became uninitialized, and zero it in debug builds.
```
Comment (Mark-Simulacrum) [inline library/alloc/src/collections/btree/node.rs:1747]: What do you mean by zero it in debug builds? This isn't zeroing anything...

**Comment by Mark-Simulacrum** [comment]
@RalfJung I am interested in hearing your thoughts on a move and move_nonoverlapping function on ptr or MaybeUninit - it seems kind of odd to have it hand-rolled in terms of the de-initializing and such in btree code (indeed I would rather reject this PR as-is; I think we should not hand roll primitives in the btree code - it's sufficiently complicated without needing to know new functions behavior).

File: library/alloc/src/collections/btree/node.rs (hunk 1 above)
Comment (ssomers) [reply library/alloc/src/collections/btree/node.rs:1747]: Probably not contractually, but in practice it zeroes, both in builds of alloc and [in this playground example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fec00574e79240334d10ee97546cab95).

**Comment by ssomers** [comment]
> I think we should not hand roll primitives in the btree code - it's sufficiently complicated without needing to know new functions behavior).

This baffles me. Do you wish `slice_insert` and `slice_remove` (which I just happened to have enhanced similarly) would not exist and presumably be written out? How about all other non-public functions in the btree realm, or the primitive data structure offered by the node module?

Naturally I wish there were standard ways for working with variable length arrays of `MaybeUninit` containers.

**Comment by Mark-Simulacrum** [comment]
I mean moving them to standard library APIs, they are of course useful as methods and I wouldn't just inline them.

File: library/alloc/src/collections/btree/node.rs [hunk 2]
```diff
@@ -1739,9 +1725,35 @@ unsafe fn slice_remove<T>(slice: &mut [MaybeUninit<T>], idx: usize) -> T {
         let slice_ptr = slice.as_mut_ptr();
         let ret = (*slice_ptr.add(idx)).assume_init_read();
         ptr::copy(slice_ptr.add(idx + 1), slice_ptr.add(idx), len - idx - 1);
+        if cfg!(debug_assertions) {
+            // Tell Miri about the trailing element, and zero it in debug builds.
```
Comment (RalfJung) [inline library/alloc/src/collections/btree/node.rs:1729]: What does it mean to "tell Miri" about an element?

Comment (ssomers) [reply library/alloc/src/collections/btree/node.rs:1729]: As in the description, that is became uninitialized.

Comment (RalfJung) [reply library/alloc/src/collections/btree/node.rs:1729]: Ah I see... that wasn't clear (also because it's in the hidden part of the diff).

**Comment by RalfJung** [comment]
> I am interested in hearing your thoughts on a move and move_nonoverlapping function on ptr or MaybeUninit 

This is a weird operation... it's not really useful aside from making Miri complain about future uses of this memory (assuming those uses actually do things that are illegal for uninit memory, such as arithmetic/logic operations). Or am I missing something?

**Comment by ssomers** [comment]
> it's not really useful aside from making Miri complain

I'm not sure you're talking about `slice_move`, but I call Miri's complaints really useful regardless. Aside from that, `slice_move` checks sizes and conveys to the reader that it transfers ownership, whereas `copy_nonoverlapping` suggests it copies.

**Comment by RalfJung** [comment]
> whereas copy_nonoverlapping suggests it copies.

I guess I considered it to work on a level below the one where "copy" vs "move" is a meaningful distinction. I view this as being mostly a type-system-level fiction; on the level of unsafe code messing around with the underlying bytes, they are both the same.

In particular, note that the exact behavior of "MIR-level `move`" is not determined yet, and whether it replaces the source by `undef` is unclear. Miri currently treats copy and move exactly identical. Also see https://github.com/rust-lang/unsafe-code-guidelines/issues/188.

> I call Miri's complaints really useful regardless

Fair. I guess what I meant to say is -- there is no precedent for APIs that express such "Miri-only effects".

**Comment by ssomers** [comment]
> > whereas copy_nonoverlapping suggests it copies.
> 
> I guess I considered it to work on a level below the one where "copy" vs "move" is a meaningful distinction.

Yes, that's another way of putting it: ptr::copy_nonoverlapping is too low level for arrays of `MaybeUninit<T>`, that really model  an `Option<T>` hiding whether it contains a value (and thus doesn't drop the value). Or at least, that's how the btree code treats them, and that's what Miri appears to track meticulously (when asked to).

Once you have that, there is some meaningful distinction between "move" and "copy". But what `slice_move` does is incomplete because it can also overwrite an initialized `MaybeUninit<T>` and you loose an owned value. I'm not going to try to complete that ownership idea, so let's say we rename `slice_move` to `slice_copy_nonoverlapping` and forget the marking of uninitialized positions. Actually, that should be `slice_copy` because if you have an exclusive `&mut [MaybeUninit<T>]` destination, the source slice cannot overlap.

What's left is the `debug_assert` on slice lengths, the most important bit that Miri's pointer provenance checking detected late in various places. One might say, that's what [copy_from_slice](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) already does, so let's try to mimic that, without the Copy bound.

> Fair. I guess what I meant to say is -- there is no precedent for APIs that express such "Miri-only effects".

Isn't [MaybeUninit::uninit](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fec00574e79240334d10ee97546cab95) a "debug build and Miri-only effect"?

**Comment by RalfJung** [comment]
> Isn't MaybeUninit::uninit a "debug build and Miri-only effect"?

No, that function is called in many situations where people don't care about Miri at all.

File: library/alloc/src/collections/btree/node.rs [hunk 3]
```diff
@@ -1741,5 +1734,14 @@ unsafe fn slice_remove<T>(slice: &mut [MaybeUninit<T>], idx: usize) -> T {
     }
 }
 
+/// Works like `dst.copy_from_slice(src)` but does not require `T` to be `Copy` nor `Clone`.
+/// Useful to implement a move, where you consider `src` uninitialized afterwards.
+fn copy_from_slice<T>(dst: &mut [MaybeUninit<T>], src: &[MaybeUninit<T>]) {
```
Comment (Mark-Simulacrum) [inline library/alloc/src/collections/btree/node.rs:1739]: I think this should be called "move_from_slice" and the argument order made consistent with copy_nonoverlapping.

Comment (ssomers) [reply library/alloc/src/collections/btree/node.rs:1739]: Do you really want the order of a `move_from_slice` to be the inverse of [slice::copy_from_slice](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice)? And should edges use `slice::copy_from_slice` (as in the current PR) or this `move_from_slice`?

Comment (Mark-Simulacrum) [reply library/alloc/src/collections/btree/node.rs]: copy_from_slice is a method, and this is a free function, so conventions/naming are a bit different imo.

move_to_slice would be great, excellent point.

**Comment by Mark-Simulacrum** [comment]
modulo rename r=me

**Pull request merged**